packaging/_tokenizer.py

```diff
@@ -0,0 +1,154 @@
+        self.name = name
+        self.text = text
+        self.position = position
+
+    def matches(self, name: str = "", text: str = "") -> bool:
+        if name and self.name != name:
+            return False
+        return True
+
+
+class ParseException(Exception):
+    """Parsing failed"""
+
+    def __init__(self, message: str, position: int) -> None:
+        super().__init__(message)
+        self.position = position
+
+
+DEFAULT_RULES = {
+    None: r"[ \t]+",  # whitespace: not returned as tokens
+    "LPAREN": r"\(",
+    "RPAREN": r"\)",
+    "LBRACKET": r"\[",
+    "RBRACKET": r"\]",
+    "SEMICOLON": r";",
+    "COLON": r",",
+    "QUOTED_STRING": re.compile(
+        r"""
+            ('[^']*')
+            |
+            ("[^"]*")
+        """,
+        re.VERBOSE,
+    ),
+    "OP": r"===|==|~=|!=|<=|>=|<|>",
+    "VERSION": re.compile(Specifier._version_regex_str, re.VERBOSE | re.IGNORECASE),
+    "BOOLOP": r"or|and",
+    "IN": r"in",
+    "NOT": r"not",
+    "VARIABLE": re.compile(
+        r"""
+            python_version
+            |python_full_version
+            |os[._]name
+            |sys[._]platform
+            |platform_(release|system)
+            |platform[._](version|machine|python_implementation)
+            |python_implementation
+            |implementation_(name|version)
+            |extra
+        """,
+        re.VERBOSE,
+    ),
+    "URL_SPEC": "@ *[^ ]+",
+    "IDENTIFIER": r"([a-zA-Z0-9]|-|_|\.)+",
+}
+
+
+class Tokenizer:
+    """Stream of tokens for a LL(1) parser.
+
+    Provides methods to examine the next token to be read, and to read it
+    (advance to the next token).
+    """
+
+    next_token: Optional[Token]
+
+    def __init__(
+        self, source: str, rules: Dict[Optional[str], object] = DEFAULT_RULES
+    ) -> None:
+        self.source = source
+        self.rules = {name: re.compile(pattern) for name, pattern in rules.items()}
+        self.next_token = None
+        self.generator = self._tokenize()
+        self.position = 0
+
+    def peek(self) -> Token:
+        """Return the next token to be read"""
+        if not self.next_token:
+            self.next_token = next(self.generator)
+        return self.next_token
+
+    def match(self, *match_args: str) -> bool:
+        """Return True if the next token matches the given arguments"""
+        token = self.peek()
+        return token.matches(*match_args)
+
+    def expect(self, *match_args: str, **match_kwargs: str) -> Token:
+        """Raise SyntaxError if the next token doesn't match given arguments"""
+        token = self.peek()
+        if not token.matches(*match_args, **match_kwargs):
+            exp = " ".join(
+                v
+                for v in match_args
+                + tuple(f"{k}={v!r}" for k, v in match_kwargs.items())
+                if v
+            )
+            raise self.raise_syntax_error(message=f"Expected {exp}")
```

Comment:
It turns out that the autogenerated  error messages aren't really used. It might be better if `expect` and `read` took an `error_message` argument.